### PR TITLE
PCountable and PEnumerable

### PR DESCRIPTION
**Pull request by kozross**
This adds two type classes, providing successor and predecessor operations. This is groundwork for fixing #705 . We still need further work on fixing `PPosixTime` to support negative values as well (as per #704), but I'll send this in a separate PR.

### CODE REVIEW

File: Plutarch/Enum.hs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import Plutarch.Internal.PLam (plam)
+import Plutarch.Positive (PPositive)
+
+{- | A notion of \'next\' value. More formally, instances of this type class are
+discrete linear orders with no maximal element.
+
+= Laws
+
+1. @x /= psuccessor x@
+2. @y < x@ @=@ @x >= psuccessor y@
```
Comment (SeungheonOh) [inline Plutarch/Enum.hs:28]: @y<x@ @=@ @psuccessor y <= x@ reads better

File: Plutarch/Enum.hs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import Plutarch.Internal.Other (pfix, pto)
+import Plutarch.Internal.PLam (plam)
+import Plutarch.Positive (PPositive)
+
+{- | A notion of \'next\' value. More formally, instances of this type class are
+discrete linear orders with no maximal element.
+
+= Laws
+
+1. @x /= psuccessor x@
```
Comment (L-as) [inline Plutarch/Enum.hs:27]: What if there’s only one element

Comment (kozross) [reply Plutarch/Enum.hs:27]: Then you can't be an instance.

File: Plutarch/Enum.hs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  -- | @since WIP
+  {-# INLINEABLE psuccessorN #-}
+  psuccessorN :: forall (s :: S). Term s (PPositive :--> a :--> a)
+  psuccessorN = phoistAcyclic $ plam $ \n x -> go n # (psuccessor # x) # 1
+    where
+      go ::
+        forall (s' :: S).
+        Term s' PPositive ->
+        Term s' (a :--> PPositive :--> a)
+      go limit = pfix #$ plam $ \self acc count ->
```
Comment (SeungheonOh) [inline Plutarch/Enum.hs:56]: This is getting recursed on script. Is this more favorable to having it generate script that just runs `psuccessor` n times? 

Latter will make script size grow linearly on the argument `n` but with smaller execution cost.

File: Plutarch/Enum.hs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  -- | @since WIP
+  {-# INLINEABLE ppredecessorN #-}
+  ppredecessorN :: forall (s :: S). Term s (PPositive :--> a :--> a)
+  ppredecessorN = phoistAcyclic $ plam $ \n x -> go n # (ppredecessor # x) # 1
+    where
+      go ::
+        forall (s' :: S).
+        Term s' PPositive ->
+        Term s' (a :--> PPositive :--> a)
+      go limit = pfix #$ plam $ \self acc count ->
```
Comment (SeungheonOh) [inline Plutarch/Enum.hs:109]: Similar observation here about recursion

File: Plutarch/Enum.hs (hunk 3 above)
Comment (SeungheonOh) [reply Plutarch/Enum.hs:56]: Maybe it's better to just not provide default instance here as many types like `PInteger` `PPositive` have much better way of dealing with `psuccessorN` than what's provided as default

**Review by SeungheonOh — APPROVED** [review]
Some questions, versions needs to be correctly appended. Rest looks good to me. You can merge once versioning is addressed

File: Plutarch/Enum.hs (hunk 3 above)
Comment (kozross) [reply Plutarch/Enum.hs:56]: Running `psuccessor` N times would require N to be a constant, which is a bit limiting. 

I reckon that a default instance, assuming we note that it's not particularly efficient, isn't a problem. All of our existing instances are hand-rolled anyway.

**Review by SeungheonOh — DISMISSED** [review]

**Review by t4ccer — APPROVED** [review]
I'm not sure if we should `phoistAcyclic` things like `psuccessor = phoistAcyclic $ plam (+)` as I get that `plam (+)` would get inlined as a single builtin otherwise, but you probably need to have benchmarks to support such claim. Good for know I guess, there was also some recent discussion on Plutonomicon discord on excessive use of `phoistAcyclic` so I guess it's better not to block over this and batch review it later.

**Pull request merged**